modules/contact/src/constraints/NormalNodalLMMechanicalContact.C

```diff
@@ -0,0 +1,204 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NormalNodalLMMechanicalContact.h"
+#include "PenetrationLocator.h"
+#include "PenetrationInfo.h"
+#include "SystemBase.h"
+#include "Assembly.h"
+
+#include "metaphysicl/dualnumberarray.h"
+
+using MetaPhysicL::DualNumber;
+using MetaPhysicL::NumberArray;
+
+registerMooseObject("MooseApp", NormalNodalLMMechanicalContact);
+
+template <>
+InputParameters
+validParams<NormalNodalLMMechanicalContact>()
+{
+  InputParameters params = validParams<NodeFaceConstraint>();
+  params.set<bool>("use_displaced_mesh") = true;
+
+  params.addCoupledVar("disp_y", "The y displacement");
+  params.addCoupledVar("disp_z", "The z displacement");
+  params.addParam<Real>("c", 1, "Parameter for balancing the size of the gap and contact pressure");
+
+  MooseEnum ncp_function_type("min fb", "min");
+  params.addParam<MooseEnum>(
+      "ncp_function_type", ncp_function_type, "The type of NCP function to use");
+
+  params.addClassDescription("Implements the KKT conditions for normal contact using an NCP "
+                             "function. Requires that either the gap distance or the normal "
+                             "contact pressure (represented by the value of `variable`) is zero.");
+  return params;
+}
+
+NormalNodalLMMechanicalContact::NormalNodalLMMechanicalContact(const InputParameters & parameters)
+  : NodeFaceConstraint(parameters),
+    _disp_y_id(coupled("disp_y")),
+    _disp_z_id(coupled("disp_z")),
+    _c(getParam<Real>("c")),
+    _epsilon(std::numeric_limits<Real>::epsilon()),
+    _ncp_type(getParam<MooseEnum>("ncp_function_type"))
+
+{
+  _overwrite_slave_residual = false;
+}
+
+Real
+NormalNodalLMMechanicalContact::computeQpSlaveValue()
+{
+  return _u_slave[_qp];
+}
+
+void
+NormalNodalLMMechanicalContact::computeResidual()
+{
+  DenseVector<Number> & re = _assembly.residualBlock(_var.number());
+
+  _qp = 0;
+
+  re(0) = computeQpResidual(Moose::Slave);
+}
```

Comment:
It looks like you are not supporting the tagging capability. It would be great, if we could add that.

Reply:
Slack pack not doing that :smile: Someday